File: ui/src/stores/keyframeStore/query.ts
/**
 * Keyframe Query Utilities
 *
 * Functions for querying keyframe state and navigation.
 */

import type { AnimatableProperty, Keyframe, LayerTransform, PropertyValue } from "@/types/project";
import type { KeyframeStoreAccess } from "./types";

// ============================================================================
// KEYFRAME QUERY UTILITIES
// ============================================================================

/**
 * Get keyframes at a specific frame across all animated properties of a layer.
 */
export function getKeyframesAtFrame(
  store: KeyframeStoreAccess,
  layerId: string,
  frame: number,
): Array<{ propertyPath: string; keyframe: Keyframe<PropertyValue> }> {
  const layer = store.getActiveCompLayers().find((l) => l.id === layerId);
  if (!layer) return [];

  const results: Array<{ propertyPath: string; keyframe: Keyframe<PropertyValue> }> = [];

  // Check transform properties
  const transformProps: Array<keyof LayerTransform> = ["position", "scale", "rotation", "anchorPoint"];
  for (const propName of transformProps) {
    const prop = layer.transform[propName] as AnimatableProperty<unknown> | undefined;
    if (prop && "animated" in prop && prop.animated && prop.keyframes) {
      const kf = prop.keyframes.find((k) => k.frame === frame);
      if (kf) {
        // Cast: keyframe values are always PropertyValue-compatible
        results.push({ propertyPath: propName, keyframe: kf as Keyframe<PropertyValue> });
      }
    }
  }

  // Check opacity
  if (layer.opacity?.animated && layer.opacity.keyframes) {
    const kf = layer.opacity.keyframes.find((k) => k.frame === frame);
    if (kf) {
      results.push({ propertyPath: "opacity", keyframe: kf });
    }
  }

  // Check 3D rotations
  const threeDProps: Array<keyof LayerTransform> = ["rotationX", "rotationY", "rotationZ", "orientation"];
  for (const propName of threeDProps) {
    const prop = layer.transform[propName] as AnimatableProperty<unknown> | undefined;
    if (prop && "animated" in prop && prop.animated && prop.keyframes) {
      const kf = prop.keyframes.find((k) => k.frame === frame);
      if (kf) {
        // Cast: keyframe values are always PropertyValue-compatible
        results.push({ propertyPath: propName, keyframe: kf as Keyframe<PropertyValue> });
      }
    }
  }

  // Check custom properties
  for (const prop of layer.properties) {
    if (prop.animated && prop.keyframes) {
      const kf = prop.keyframes.find((k) => k.frame === frame);
      if (kf) {
        results.push({ propertyPath: prop.name, keyframe: kf });
      }
    }
  }

  return results;
}

/**
 * Get all keyframe frames for a layer (for timeline display).
 */
export function getAllKeyframeFrames(
  store: KeyframeStoreAccess,
  layerId: string,
): number[] {
  const layer = store.getActiveCompLayers().find((l) => l.id === layerId);
  if (!layer) return [];

  const frames = new Set<number>();

  // Collect frames from transform properties
  const transformProps: Array<keyof LayerTransform> = ["position", "scale", "rotation", "anchorPoint"];
  for (const propName of transformProps) {
    const prop = layer.transform[propName] as AnimatableProperty<unknown> | undefined;
    if (prop && "animated" in prop && prop.animated && prop.keyframes) {
      for (const kf of prop.keyframes) {
        frames.add(kf.frame);
      }
    }
  }

  // Collect from opacity
  if (layer.opacity?.animated && layer.opacity.keyframes) {
    for (const kf of layer.opacity.keyframes) {
      frames.add(kf.frame);
    }
  }

  // Collect from 3D properties
  const threeDProps: Array<keyof LayerTransform> = ["rotationX", "rotationY", "rotationZ", "orientation"];
  for (const propName of threeDProps) {
    const prop = layer.transform[propName] as AnimatableProperty<unknown> | undefined;
    if (prop && "animated" in prop && prop.animated && prop.keyframes) {
      for (const kf of prop.keyframes) {
        frames.add(kf.frame);
      }
    }
  }

  // Collect from custom properties
  for (const prop of layer.properties) {
    if (prop.animated && prop.keyframes) {
      for (const kf of prop.keyframes) {
        frames.add(kf.frame);
      }
    }
  }

  return Array.from(frames).sort((a, b) => a - b);
}

/**
 * Find the next keyframe frame after the given frame.
 *
 * @param store - The keyframe store
 * @param currentFrame - The current frame
 * @param layerIds - Layer IDs to search (if empty, searches all layers)
 * @returns The next keyframe frame, or null if none found
 */
export function findNextKeyframeFrame(
  store: KeyframeStoreAccess,
  currentFrame: number,
  layerIds: string[],
): number | null {
  const layers = store.getActiveCompLayers();
  const searchLayerIds =
    layerIds.length > 0 ? layerIds : layers.map((l) => l.id);

  const frameSet = new Set<number>();
  for (const lid of searchLayerIds) {
    for (const frame of getAllKeyframeFrames(store, lid)) {
      frameSet.add(frame);
    }
  }

  const allFrames = Array.from(frameSet).sort((a, b) => a - b);
  return allFrames.find((f) => f > currentFrame) ?? null;
}

/**
 * Find the previous keyframe frame before the given frame.
 *
 * @param store - The keyframe store
 * @param currentFrame - The current frame
 * @param layerIds - Layer IDs to search (if empty, searches all layers)
 * @returns The previous keyframe frame, or null if none found
 */
export function findPrevKeyframeFrame(
  store: KeyframeStoreAccess,
  currentFrame: number,
  layerIds: string[],
): number | null {
  const layers = store.getActiveCompLayers();
  const searchLayerIds =
    layerIds.length > 0 ? layerIds : layers.map((l) => l.id);

  const frameSet = new Set<number>();
  for (const lid of searchLayerIds) {
    for (const frame of getAllKeyframeFrames(store, lid)) {
      frameSet.add(frame);
    }
  }

  const allFrames = Array.from(frameSet).sort((a, b) => a - b);
  const prevFrames = allFrames.filter((f) => f < currentFrame);
  return prevFrames.length > 0 ? prevFrames[prevFrames.length - 1] : null;
}

/**
 * Find the nearest keyframes before and after a given frame.
 */
export function findSurroundingKeyframes<T>(
  property: AnimatableProperty<T>,
  frame: number,
): { before: Keyframe<T> | null; after: Keyframe<T> | null } {
  if (!property.keyframes || property.keyframes.length === 0) {
    return { before: null, after: null };
  }

  let before: Keyframe<T> | null = null;
  let after: Keyframe<T> | null = null;

  for (const kf of property.keyframes) {
    if (kf.frame <= frame) {
      before = kf;
    } else if (kf.frame > frame && after === null) {
      after = kf;
      break;
    }
  }

  return { before, after };
}

File: comfyui/src/lattice/State/Keyframe/Query.hs
-- |
-- Module      : Lattice.State.Keyframe.Query
-- Description : Keyframe query utilities
--
-- Migrated from ui/src/stores/keyframeStore/query.ts
-- Pure functions for querying keyframe state and navigation
--

{-# LANGUAGE OverloadedStrings #-}

module Lattice.State.Keyframe.Query
  ( getKeyframesAtFrame
  , getAllKeyframeFrames
  , findNextKeyframeFrame
  , findPrevKeyframeFrame
  , findSurroundingKeyframes
  -- Types
  , KeyframeAtFrame(..)
  ) where

import Data.List (find, sort, nub, reverse)
import Data.Maybe (Maybe(..))
import Data.Text (Text)
import qualified Data.Text as T
import Lattice.State.Keyframe.Helpers (findPropertyByPath)
import Lattice.State.Layer.Queries (getLayerById)
import Lattice.Types.Animation
  ( AnimatableProperty(..)
  , Keyframe(..)
  , PropertyValue(..)
  )
import Lattice.Types.Layer (Layer(..))
import Lattice.Types.Primitives (validateFinite)

-- ============================================================================
-- TYPES
-- ============================================================================

-- | Keyframe at a specific frame with property path
data KeyframeAtFrame = KeyframeAtFrame
  { keyframeAtFramePropertyPath :: Text
  , keyframeAtFrameKeyframe :: Keyframe PropertyValue
  }
  deriving (Eq, Show)

-- ============================================================================
-- KEYFRAME QUERY UTILITIES
-- ============================================================================

-- | Get keyframes at a specific frame across all animated properties of a layer
-- Pure function: takes layer ID, frame, and layers list
-- Returns list of keyframes with their property paths
getKeyframesAtFrame ::
  Text -> -- Layer ID
  Double -> -- Frame
  [Layer] -> -- Layers list
  [KeyframeAtFrame] -- List of keyframes at frame with property paths
getKeyframesAtFrame targetLayerId frame layers =
  case getLayerById layers targetLayerId of
    Nothing -> []
    Just layer ->
      let
        -- Validate frame is finite and >= 1
        safeFrame = if validateFinite frame && frame >= 1 then frame else 1.0
        
        -- Property paths to check
        propertyPaths =
          [ "position"
          , "scale"
          , "rotation"
          , "origin"
          , "anchorPoint"
          , "rotationX"
          , "rotationY"
          , "rotationZ"
          , "orientation"
          , "opacity"
          ]
        
        -- Collect keyframes from standard properties using findPropertyByPath
        standardKeyframes = concatMap (\propPath ->
          case findPropertyByPath layer propPath of
            Nothing -> []
            Just prop ->
              if animatablePropertyAnimated prop && not (null (animatablePropertyKeyframes prop))
                then
                  case find (\k -> keyframeFrame k == safeFrame) (animatablePropertyKeyframes prop) of
                    Nothing -> []
                    Just kf -> [KeyframeAtFrame propPath kf]
                else []
          ) propertyPaths
        
        -- Collect from custom properties (layerProperties is [Value] until migrated to [AnimatableProperty PropertyValue])
        customKeyframes = []
      in
        standardKeyframes ++ customKeyframes

-- | Get all keyframe frames for a layer (for timeline display)
-- Pure function: takes layer ID and layers list
-- Returns sorted list of unique frame numbers
getAllKeyframeFrames ::
  Text -> -- Layer ID
  [Layer] -> -- Layers list
  [Double] -- Sorted list of unique frame numbers
getAllKeyframeFrames targetLayerId layers =
  case getLayerById layers targetLayerId of
    Nothing -> []
    Just layer ->
      let
        -- Property paths to check
        propertyPaths =
          [ "position"
          , "scale"
          , "rotation"
          , "origin"
          , "anchorPoint"
          , "rotationX"
          , "rotationY"
          , "rotationZ"
          , "orientation"
          , "opacity"
          ]
        
        -- Collect frames from standard properties using findPropertyByPath
        standardFrames = concatMap (\propPath ->
          case findPropertyByPath layer propPath of
            Nothing -> []
            Just prop ->
              if animatablePropertyAnimated prop && not (null (animatablePropertyKeyframes prop))
                then map keyframeFrame (animatablePropertyKeyframes prop)
                else []
          ) propertyPaths
        
        -- Collect from custom properties
        customFrames = concatMap (\prop ->
          if animatablePropertyAnimated prop && not (null (animatablePropertyKeyframes prop))
            then map keyframeFrame (animatablePropertyKeyframes prop)
            else []
          ) (layerProperties layer)
        
        -- Combine, remove duplicates, sort
        allFrames = nub (standardFrames ++ customFrames)
      in
        sort allFrames

-- | Find the next keyframe frame after the given frame
-- Pure function: takes current frame, layer IDs, and layers list
-- Returns next keyframe frame or error
findNextKeyframeFrame ::
  Double -> -- Current frame
  [Text] -> -- Layer IDs to search (empty = search all layers)
  [Layer] -> -- Layers list
  Either Text Double -- Next keyframe frame or error
findNextKeyframeFrame currentFrame layerIds layers =
  let
    -- Validate current frame
    safeCurrentFrame = if validateFinite currentFrame && currentFrame >= 1 then currentFrame else 1.0
    
    -- Determine which layers to search
    searchLayerIds = if null layerIds
      then map layerId layers
      else layerIds
    
    -- Collect all frames from all search layers
    allFramesSet = foldl (\acc lid -> acc ++ getAllKeyframeFrames lid layers) [] searchLayerIds
    uniqueFrames = sort (nub allFramesSet)
    
    -- Find next frame
    mNextFrame = find (\f -> f > safeCurrentFrame) uniqueFrames
  in
    case mNextFrame of
      Nothing -> Left ("No next keyframe found after frame " <> T.pack (show safeCurrentFrame) <> " for layers: " <> T.intercalate ", " searchLayerIds)
      Just nextFrame -> Right nextFrame

-- | Find the previous keyframe frame before the given frame
-- Pure function: takes current frame, layer IDs, and layers list
-- Returns previous keyframe frame or error
findPrevKeyframeFrame ::
  Double -> -- Current frame
  [Text] -> -- Layer IDs to search (empty = search all layers)
  [Layer] -> -- Layers list
  Either Text Double -- Previous keyframe frame or error
findPrevKeyframeFrame currentFrame layerIds layers =
  let
    -- Validate current frame
    safeCurrentFrame = if validateFinite currentFrame && currentFrame >= 1 then currentFrame else 1.0
    
    -- Determine which layers to search
    searchLayerIds = if null layerIds
      then map layerId layers
      else layerIds
    
    -- Collect all frames from all search layers
    allFramesSet = foldl (\acc lid -> acc ++ getAllKeyframeFrames lid layers) [] searchLayerIds
    uniqueFrames = sort (nub allFramesSet)
    
    -- Find previous frames (before current)
    prevFrames = filter (\f -> f < safeCurrentFrame) uniqueFrames
  in
    case prevFrames of
      [] -> Left ("No previous keyframe found before frame " <> T.pack (show safeCurrentFrame) <> " for layers: " <> T.intercalate ", " searchLayerIds)
      _ ->
        let
          reversed = reverse prevFrames
          mLastFrame = case reversed of
            [] -> Nothing
            (f : _) -> Just f
        in
          case mLastFrame of
            Nothing -> Left ("No previous keyframe found before frame " <> T.pack (show safeCurrentFrame) <> " for layers: " <> T.intercalate ", " searchLayerIds)
            Just lastFrame -> Right lastFrame

-- | Find the nearest keyframes before and after a given frame
-- Pure function: takes property and frame
-- Returns before and after keyframes (Nothing if not found)
findSurroundingKeyframes ::
  AnimatableProperty a -> -- Property to search
  Double -> -- Frame
  (Maybe (Keyframe a), Maybe (Keyframe a)) -- (before, after) keyframes
findSurroundingKeyframes property frame =
  let
    -- Validate frame
    safeFrame = if validateFinite frame && frame >= 1 then frame else 1.0
    keyframes = animatablePropertyKeyframes property
  in
    if null keyframes
      then (Nothing, Nothing)
      else
        let
          -- Find before and after
          (before, after) = findSurroundingKeyframes' keyframes safeFrame Nothing Nothing
        in
          (before, after)
  where
    findSurroundingKeyframes' ::
      [Keyframe a] ->
      Double ->
      Maybe (Keyframe a) ->
      Maybe (Keyframe a) ->
      (Maybe (Keyframe a), Maybe (Keyframe a))
    findSurroundingKeyframes' [] _ before after = (before, after)
    findSurroundingKeyframes' (kf : kfs) f before after =
      if keyframeFrame kf <= f
        then findSurroundingKeyframes' kfs f (Just kf) after
        else
          case after of
            Nothing -> (before, Just kf)
            Just _ -> (before, after)
